packages/ember-routing/lib/system/route.js

```diff
@@ -92,23 +99,76 @@ var Route = EmberObject.extend(ActionHandler, Evented, {
     @property _qp
   */
   _qp: computed(function() {
-    var controllerName = this.controllerName || this.routeName;
-    var controllerClass = this.container.lookupFactory(`controller:${controllerName}`);
+    var controllerProto, combinedQueryParameterConfiguration;
 
-    if (!controllerClass) {
-      return defaultQPMeta;
+    var controllerName = this.controllerName || this.routeName;
+    var definedControllerClass = this.container.lookupFactory(`controller:${controllerName}`);
+    var queryParameterConfiguraton = get(this, 'queryParams');
+
+    if (definedControllerClass) {
+      // the developer has authored a controller class in their application for this route
+      // access the prototype, find its query params and normalize their object shape
+      // them merge in the query params for the route. As a mergedProperty, Route#queryParams is always
+      // at least `{}`
+      controllerProto = definedControllerClass.proto();
+
+      var controllerDefinedQueryParameterConfiguration = get(controllerProto, 'queryParams');
+      var normalizedControllerQueryParameterConfiguration = normalizeControllerQueryParams(controllerDefinedQueryParameterConfiguration);
+
+      combinedQueryParameterConfiguration = mergeEachQueryParams(normalizedControllerQueryParameterConfiguration, queryParameterConfiguraton);
+
+    } else if (keys(queryParameterConfiguraton).length) {
```

Comment:
As a mergedProperty, Route#queryParams is always at least `{}`, so `queryParameterConfiguraton` is always defined.